core/filelocation.cpp

```diff
@@ -0,0 +1,235 @@
+{
+}
+
+bool isRemoteRepository(const QString &name)
+{
+	return name.startsWith("http://") || name.startsWith("https://") ||
+	       name.startsWith("ssh://") || name.startsWith("git://");
+}
+
+// For git repositories, remove "file://" and extract username, if any.
+// For local files or git repositories, do proper unicode encoding.
+void FileLocation::fixName()
+{
+	switch (type) {
+	default:
+	case NONE:
+		return;
+	case LOCAL_FILE:
+		name = QFile::encodeName(QDir::toNativeSeparators(name));
+		return;
+	case CLOUD_GIT:
+	case CLOUD_GIT_OFFLINE:
+	case GIT:
+		if (name.startsWith("file://", Qt::CaseInsensitive)) {
+			// Treat "file://..." URLs as local repositories.
+			name = name.mid(7);
+			name = QFile::encodeName(QDir::toNativeSeparators(name));
+		} else if(!isRemoteRepository(name)) {
+			name = QFile::encodeName(QDir::toNativeSeparators(name));
+		} else if(name.startsWith("https://", Qt::CaseInsensitive)) {
+			int at = name.indexOf('@');
+			int slash = name.indexOf('/', 8);
+			if (at >= 0 && slash >= 0 && slash > at) {
+				user = name.mid(8, at - 8);
+				name.remove(8, at - 7);
+			}
+		}
+	}
+}
+
+// Check if filename is of the form "repository[branch]" or "repository[branch].ssrf".
+static bool isGitFilename(const QString &name)
+{
+	QString trimmed = name.trimmed();
+	if (trimmed.endsWith(".ssrf", Qt::CaseInsensitive))
+		trimmed.truncate(trimmed.count() - 5);
+	return trimmed.right(1) == "]" &&
+	       trimmed.lastIndexOf('[') > 0;
+}
+
+// Returns (repository, branch) pair.
+static std::pair<QString, QString> parseGitFilename(const QString &fn)
+{
+	if (!isGitFilename(fn))							// Oops. Filename doesn't follow the rules. Leave branch empty.
+		return std::make_pair(fn, QString());
+
+	int branch_from = fn.lastIndexOf('[');
+	int branch_to = fn.lastIndexOf(']');
+	QString repository = fn.left(branch_from);				// Guaranteed OK owing to isGitFilename() == true
+	QString branch = fn.mid(branch_from + 1, branch_to - branch_from - 1);	// ditto
+	return std::make_pair(repository, branch);
+}
+
+FileLocation::FileLocation(Type type_, const QString &name_, const QString &branch_) : type(type_), name(name_), branch(branch_)
+{
+	if (type == NONE)
+		return;
+	fixName();
+}
+
+FileLocation::FileLocation(Type type_, const QString &name_) : type(type_)
+{
+	if (type == NONE)
+		return;
+	if (type == LOCAL_FILE) {
+		name = name_;
+	} else {
+		// Is a git repository
+		std::pair<QString, QString> repo_branch = parseGitFilename(name_);
+		name = repo_branch.first;
+		branch = repo_branch.second;
+	}
+	fixName();
+}
+
+FileLocation::FileLocation(const QString &type_, const QString &name_, const QString &branch_) : name(name_), branch(branch_)
+{
+	if (type_ == "none")
+		type = NONE;
+	else if (type_ == "local")
+		type = LOCAL_FILE;
+	else if (type_ == "cloud")
+		type = CLOUD_GIT;
+	else if (type_ == "cloud_offline")
+		type = CLOUD_GIT_OFFLINE;
+	else if (type_ == "git")
+		type = GIT;
+	else
```

Comment:
I'm surprised about the three different types of git storage - I guess I'll need to read more of the code to truly understand what you are doing here

Reply:
Note that I'm just encoding existing practice here and not changing any logic.

The existing code handles these three cases differently and was differentiating between these things based on the filename and the global `prefs.git_local_only flag` (which was then manipulated by the git code - oh my). The new code makes the difference explicit and makes it unnecessary to save the `git_local_only` flag, because it might change. So you it reads `if (git_state->is_cloud)` instead of `if (strstr(remote, prefs.cloud_git_url))`

Yes, it might be a good idea to unify these three cases in the future - but at the moment the code is what it is.

Reply:
That seems reasonable. I certainly agree that this will be much easier to read at the end.
As I just said in my other comment - introducing the new helpers / new structures in their own commit, and then replacing the use cases one at a time might make it a lot easier to understand the changes.